File: python/array/FirstMissingPositive.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def firstMissingPositive(self, A):
        i = 0
        while i < len(A):
            """
            The last condition is used to deal with the second test case.
            i = 3, A[i] = 2
            A[i]-1 = 1
            A[A[i]-1] = 2
            we should skip this one. as well as i = 4
            """
            if i + 1 == A[i] or A[i] > len(A) or A[i] <= 0 or A[A[i]-1] == A[i]:
                i += 1
            else:
                temp = A[i]
                target = A[i] - 1
                A[i] = A[target]
                A[target] = temp
            

        meet_one = False
        for i in xrange(len(A)):
            if not meet_one and A[i] == 1:
                meet_one = True
            if meet_one and A[i] != i+1:
                return i+1

        if not meet_one:
            return 1

        return len(A) + 1

if __name__ == "__main__":
    solution = Solution()

    #print solution.firstMissingPositive([2, 3, 5, 1])
    print solution.firstMissingPositive([1, 2, 2, 2])

File: c++/array/FirstMissingPositive.cpp
class Solution {
public:
    int firstMissingPositive(int A[], int n) {
        int i = 0;
        while (i < n) {
            if (i+1 == A[i] || A[i] >= n || A[i] < 0 || A[A[i]-1] == A[i]) {
                i ++;
            } else {
                int temp = A[i];
                int target = A[i] - 1;
                A[i] = A[target];
                A[target] = temp;
            }
        }
        
        for (int i = 0; i < n; i ++) {
            if (A[i] != i + 1) {
                return i + 1;
            }
        }
        
        return n + 1;
    }
};
